extruct/_extruct.py

```diff
@@ -38,48 +40,63 @@ def extract(htmlstring, base_url=None, encoding="UTF-8",
-        processors.append(('microformat', MicroformatExtractor().extract_items, htmlstring))
-    if 'rdfa' in syntaxes:
-        processors.append(('rdfa', RDFaExtractor().extract_items, tree))
-    output = {}
-    for label, extract, document in processors:
-        try:
-            output[label] = list(extract(document, base_url=base_url))
-        except Exception:
-            if errors == 'log':
-                logger.exception('Failed to extract {}'.format(label))
-            if errors == 'ignore':
-                pass
-            if errors == 'strict':
-                raise
-
-    if uniform:
+    try:
+        tree = parse_xmldom_html(htmlstring, encoding=encoding)
+        processors = []
         if 'microdata' in syntaxes:
-            output['microdata'] = _umicrodata_microformat(output['microdata'],
-                                                          schema_context=schema_context)
-        if 'microformat' in syntaxes:
-            output['microformat'] = _umicrodata_microformat(output['microformat'],
-                                                            schema_context='http://microformats.org/wiki/')
+            processors.append(
+                ('microdata', MicrodataExtractor(
+                    add_html_node=return_html_node).extract_items, tree))
+        if 'json-ld' in syntaxes:
+            processors.append(('json-ld', JsonLdExtractor().extract_items,
+                               tree))
         if 'opengraph' in syntaxes:
-            output['opengraph'] = _uopengraph(output['opengraph'])
+            processors.append(('opengraph', OpenGraphExtractor().extract_items,
+                               tree))
+        if 'microformat' in syntaxes:
+            processors.append(
+                ('microformat', MicroformatExtractor().extract_items,
+                 htmlstring))
+        if 'rdfa' in syntaxes:
+            processors.append(('rdfa', RDFaExtractor().extract_items, tree))
+        output = {}
+        for label, extract, document in processors:
+            try:
+                output[label] = list(extract(document, base_url=base_url))
+            except Exception:
+                if errors == 'log':
+                    logger.exception('Failed to extract {}'.format(label))
+                if errors == 'ignore':
+                    pass
+                if errors == 'strict':
+                    raise
+
+        if uniform:
+            if 'microdata' in syntaxes:
+                output['microdata'] = _umicrodata_microformat(
+                    output['microdata'], schema_context=schema_context)
+            if 'microformat' in syntaxes:
+                output['microformat'] = _umicrodata_microformat(
+                    output['microformat'],
+                    schema_context='http://microformats.org/wiki/')
+            if 'opengraph' in syntaxes:
+                output['opengraph'] = _uopengraph(output['opengraph'])
+    except Exception as e:
```

Comment:
If uniform fails for only one syntax shouldn't we just set that to None and return the others? This would make the code a bit cluttered since I have to wrap every case in a try, except but I feel it makes more sense

Reply:
Yes, that would be better, also I think it's possible to avoid repeating exception handling code for each unifyer in a similar way to how syntaxes are handled with a list of ``processors``. But also I think it's fine to not do it in this PR and treat as a possible future improvement.